### DISCUSSION TITLE
chapter_sorting/counting_sort/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#428] [Announcements]
# chapter_sorting/counting_sort/

一本动画图解、能运行、可提问的数据结构与算法入门书

https://www.hello-algo.com/chapter_sorting/counting_sort/

**Comment by Asce-ops** [comment]
计数排序的时候最好把最小元素也先确定，这样只需要创建一个容量为 maxmimuu - minimum + 1 的计数数组 counter，counter[num - minimum] 表示元素 num 在待排序数组中出现的次数，这样既可以减小空间复杂度也可以直接推广到存在负整数的数组

**Comment by tstanon** [comment]
C语言实现计数排序完整版
```c
void counting_sort(int *arr, int size){
    if (size <= 0) return; int max_val = arr[0]; for (int i = 1; i < size; i++) { if (arr[i] > max_val) max_val = arr[i]; } // 找到最大值，用于确定计数数组大小[0, max_val]
    int *counter = (int *) calloc(max_val + 1, sizeof(int)), *res = (int *)malloc(sizeof(int) * size); // 计数数组（最终存储前缀和）；结果数组
    for (int i = 0; i < size; i++) { counter[arr[i]]++; } for (int i = 1; i <= max_val; ++i) { counter[i] += counter[i - 1]; } // 计数和计算前缀和
    for (int i = size - 1; i >= 0; i--) { int cur_val = arr[i]; res[counter[cur_val] - 1] = cur_val; counter[cur_val]--; } // 从后往前遍历原数组，通过前缀和找到元素在结果数组中的位置，存入后，前缀和计数减1
    memcpy(arr, res, sizeof(int) * size); free(counter); free(res); 
}
```
有个小问题，为什么作者给出的完整代码（第二部分）中counter数组的大小只需要m而不是m+1了呢？

**Comment by checkma-xh** [comment]
```Python
def sort(nums):
    cnt_array = [0] * (max(nums) + 1)
    new_nums = []
    
    for num in nums:
        cnt_array[num] += 1
    
    for idx, num in enumerate(cnt_array):
        new_nums.extend([idx] * num)
    
    return new_nums

nums = [1,3,32,4,1,23,17,23,6,2,9]
print(sort(nums))
```

**Comment by jasonxuzh** [comment]
最好能补充一个用完整实现排序对象数组的例子

**Comment by FrierenGreatDemonLord** [comment]
AI 给出的前缀和式的计数排序，并通过商品数量给商品名称排序的示例：
```
class Product:
    def __init__(self, name, quantity):
        self.name = name
        self.quantity = quantity

def counting_sort(products):
    if len(products) <= 1:
        return [product.name for product in products]

    # 找到数量的最大值
    max_quantity = max(product.quantity for product in products)

    # 创建计数数组
    count = [0] * (max_quantity + 1)

    # 统计每个数量的出现次数
    for product in products:
        count[product.quantity] += 1

    # 计算前缀和
    for i in range(1, len(count)):
        count[i] += count[i - 1]

    # 创建输出数组
    sorted_products = [None] * len(products)

    # 反向填充输出数组以保持稳定性
    for product in reversed(products):
        sorted_products[count[product.quantity] - 1] = product.name
        count[product.quantity] -= 1

    return sorted_products


if __name__ == "__main__":
    products = [
        Product("Apple", 5),
        Product("Banana", 3),
        Product("Orange", 3),
        Product("Grape", 2),
        Product("Cherry", 5)
    ]

    # 排序并打印结果
    sorted_names = counting_sort(products)
    print(sorted_names)
```

一句话总结：
**在计数排序的基础上通过`prefix[num] - 1` 得到对应的索引，然后将输入数组中的商品名填入到这个索引处即可**。即更改倒序遍历输入数组，通过前缀和计算索引并填入元素处的代码即可。

**Reply by FrierenGreatDemonLord** [reply]
力扣类似题目可参见：
[2418. 按身高排序](https://leetcode.cn/problems/sort-the-people/description/)（按照身高，对姓名进行排序）

**Comment by FrierenGreatDemonLord** [comment]
C 也可利用前缀运算符简化利用前缀和计算出索引并回填出的代码：
``` C
for (i = n - 1; i >= 0; i--) {
	val = arr[i];
	ret[--count[val]] = val;
}
```

**Comment by lastVigo** [comment]
前缀和——很巧妙的设计。

**Comment by xiaoyaozizaiwoweidu** [comment]
#### 包含负数的情况，通过简单映射函数来处理负数
比如：nums = [2, 3, 1, 0, 2, 2, -4, -1, 3, 0]
nums[i] > nums[j]，显然 nums[i] + 4 > nums[j] + 4 也成立，所以映射函数 f(i) = nums[i] + 4 是可以作为计数排序的映射函数的。
```python
def counting_sort(nums):
    # 找到列表中的最小值
    min_num = min(nums)

    # 计算偏移量，使最小值映射到索引 0
    offset = -min_num

    # 计算偏移后的最大值，以确定计数数组的大小
    m = max(nums) - min_num

    # 初始化计数数组，所有元素设为 0
    counter = [0] * (m + 1)

    # 统计每个数字的出现次数
    for num in nums:
        counter[num + offset] += 1

    # 计算计数数组的前缀和
    for i in range(m):
        counter[i + 1] += counter[i]

    # 初始化结果数组
    n = len(nums)
    res = [0] * n

    # 倒序遍历列表，以保持排序的稳定性
    for i in range(n - 1, -1, -1):
        num = nums[i]
        res[counter[num + offset] - 1] = num
        counter[num + offset] -= 1

    # 将排序结果复制回原始列表
    for i in range(n):
        nums[i] = res[i]


if __name__ == '__main__':
    nums = [2, 3, 1, 0, 2, 2, -4, -1, 3, 0]
    counting_sort(nums)
    print(nums)  # 期望输出: [-4, -1, 0, 0, 1, 2, 2, 2, 3, 3]
```

**Reply by jiuerya7** [reply]
nb

**Comment by Kokhba** [comment]
这个如果输入数据是对象，上述步骤 3. 就失效了能解释一下吗？
还有prefix[num] - 1为什么 代表元素 num 在结果数组 res 中最后一次出现的索引

**Reply by xbsheng** [reply]
11.9.2   完整实现中讲解的就是针对数据是对象的情况吧~

**Reply by apricity-6** [reply]
因为第一个算法的第3步填充回原数组的值不是取的原数组的值，而是counter数组的索引。等于这有个前置关系，你的原数组都是正整数，我可以建立一个counter数组，这个counter数组的索引就是你原数组里的值，counter数组里的值，是你原数组内等于counter索引的值的数量。
比如: 原数组 [4,4,2,2,2,1],counter数组: [0,1,3,0,2]。
这样你将值从counter数组写回nums是有关联关系在的，遍历counter数组，当前索引即是应该写回nums的值。[0个0，1个1，3个2，0个3，2个4]。
而如果是对象数组的话，这个关系就变得复杂了，你可以让对象内某个正整数的值作为排序条件，你就不能简单的用count数组的索引去写回值，而是需要记录原数组的索引，这样可以根据索引取到原值，res[ri]=nums[counter[ci]];

所以对于对象的基数排序解决问题就是如何找到ci和ri的关联关系，以及ci与nums数组内元素值的对应关系。所以使用了前缀和。

这个前缀和数组是基于计数数组产生的，因为根据上面的关联关系counter数组保证了该计数数组的索引跟nums内数据的值有关系，而counter数组本身记录了数据的顺序和出现的次数。
那根据顺序和次数你就能计算出该值在结果数组中的最后一个下标。
比如：[0,1,3,0,2] => [0个0，1个1，3个2，0个3，2个4] => [0,1,4,4,6]=>[ 前面0个元素[0]，前面有1个元素[0,1],前面有4个元素[0,1,2],前面有4个元素[0,1,2,3]，前面有6个元素[0,1,2,3,4] ]。

[0,1,4,4,6]
这就能得到结果有序数组的第6个元素内的值应该是4，也就是说num=4的下标应该是 6-1 = [5]，减1是因为序数是从1开始，而数组的索引从0开始，0-based。反过来，prefix[4]-1=6-1=5，这个4是原数组的值，被关联为计数数组的索引=关联为前缀和数组的索引。这个6是4前面(包含自身)总共有6个值，而4是你从原数组内扫描得到的，所以数组内肯定有一个4，所以前6个元素最后一个肯定是4。所以结果数组内值为4的最后一个元素的下标为5。

第4个元素应该是2，下标是4-1=[3]。反过来说，prefix[2]-1=4-1=3。这个4是2前面(包含自身)总共有4个值，所以前4个元素最后一个是2。所以结果数组内值为2的最后一个元素的下标为3。
prefix下标为3的值虽然也是4，但是排序过程是循环从原数组内根据值和该值在prefix数组的前缀和来获取在结果有序数组的下标，原数组没有3，所以不会有 prefix[num=3]-1=4-1=3 这种跟2冲突的情况。

**Comment by loskyertt** [comment]
```C++
#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

/* 计数排序 */
void count_sort(vector<int> &nums) {
  int max_val = INT_MIN;
  int min_val = INT_MAX;
  for (int num : nums) {
    max_val = max(max_val, num);
    min_val = min(min_val, num);
  }

  int n = max_val - min_val + 1;
  vector<int> count(n, 0);

  for (int num : nums) {
    count[num - min_val]++;
  }

  cout << "排好序的数组为：";
  for (int i = 0; i < n; i++) {
    while (count[i]) {
      cout << i + min_val << " ";
      count[i]--;
    }
  }

  cout << endl;
}

void test() {
  vector<int> nums = {2,
                      5,
                      7,
                      1, 4, 6, -2, -4, -1, 10, 10, 8, 9, 14, 11};

  count_sort(nums);
}

int main() {
  test();

  return 0;
}
```

这个怎么样？我突然想到的，也能拿来排序含负数和重复数字的情况🤔

**Comment by sunshinesDL** [comment]
K神，11.9.2内容中关于前缀和的定义：“顾名思义，索引 i 处的前缀和 Prefix[i] 等于数组前 i 个元素之和。”是否有些歧义呢？
例如，Prefix[0] 实际等于数组前1个元素之和，而非前0个元素之和，毕竟字面意思上的前0个就是没有元素，
这样理解正确吗？😄

**Reply by jiuerya7** [reply]
就是斐波那契数列

**Reply by xbsheng** [reply]
`斐波那契数列` 和 `前缀和` 不一样，`斐波那契数列` 只是前两项的和

**Comment by xiaozhuge123** [comment]
day09

**Comment by jiuerya7** [comment]
我感觉完全可以返回新数组res，这样少遍历一次数组😁

**Reply by leiddx** [reply]
填充原数组能确保函数调用完成以后释放新数组res所占内存。

**Comment by Xianhuii** [comment]
前缀和的作用：记录数字在排序后数组的位置。
例如：
1. 计数数组counter：[3,2,3,0,2]，表示0（3个）、1（2个）、2（3个）、3（0个）、4（2个）。
2. 逻辑上将计数数组转化成排序后的数组：[0,0,0,1,1,2,2,2,4,4]。
3. 前缀和：[3,5,8,8,10]，表示数字在排序后数组的最大索引地址。
4. 反向遍历数组进行填充，实际上是将数字依次按照最大索引地址填充到排序后的位置。

**Comment by SummerNi8531** [comment]
什么是前缀和：前缀和表示“小于等于当前数字的元素总数”

**Reply by CheShiping** [reply]
在计算完前缀和之后，counter[num] 表示的是：所有 ≤ num 的元素总共有多少个。

所以，num 这个值在最终排序结果中，最后一个（最右边）出现的位置就是第 counter[num] 个元素（从1开始数），对应数组下标就是 counter[num] - 1（0-based）。

🔍 举个具体例子
假设原始数组：

c
编辑
```python
nums = [2, 5, 3, 0, 2, 3, 0, 3]
```
第一步：统计频次
text
编辑
```text
数字: 0 1 2 3 4 5
次数: 2 0 2 3 0 1
```
第二步：计算前缀和（累加）
text
编辑
```text
counter[0] = 2          // ≤0 的有 2 个
counter[1] = 2 + 0 = 2  // ≤1 的有 2 个
counter[2] = 2 + 2 = 4  // ≤2 的有 4 个
counter[3] = 4 + 3 = 7  // ≤3 的有 7 个
counter[4] = 7 + 0 = 7
counter[5] = 7 + 1 = 8  // ≤5 的有 8 个（全部）
```
所以前缀和后的 counter 数组是：

text
编辑
```text
counter = [2, 2, 4, 7, 7, 8]
           ↑  ↑  ↑  ↑  ↑  ↑
          num=0 1  2  3  4  5
```
✅ 现在看几个具体数字：
1. 对于 num = 0
counter[0] = 2
意思是：≤0 的数有 2 个 → 所以两个 0 应该放在结果的 第1位和第2位（1-based）
那么最后一个 0 出现在 第2位 → 数组下标 = 2 - 1 = 1
✅ 所以 counter[0] - 1 = 1 就是 0 最后一次出现的索引
2. 对于 num = 2
counter[2] = 4 → ≤2 的数有 4 个
所以前4个位置是 [0, 0, 2, 2]
最后一个 2 在第4位 → 下标 = 4 - 1 = 3
✅ counter[2] - 1 = 3 正确！
3. 对于 num = 3
counter[3] = 7 → ≤3 的有 7 个
所以前7个位置包含所有 0,2,3
最后一个 3 在第7位 → 下标 = 6
✅ counter[3] - 1 = 6

**Comment by Oso-Kuan** [comment]
接下来，我们遍历 counter ，根据各数字*出现次数从小到大*的顺序填入 nums 即可。最后一句应该改为对应图中的文字：将各数字按照出现次数填入nums即可。

**Comment by Jacksonm75** [comment]
```c
    // 2. 统计各数字的出现次数
    // counter[num] 代表 num 的出现次数
    int *counter = calloc(m, sizeof(int));  //这里应该是m+1吧，否则会数组越界
    for (int i = 0; i < size; i++) {
        counter[nums[i]]++;
    }
```
